locator/network_topology_strategy.cc

```diff
@@ -307,42 +307,70 @@ effective_replication_map_ptr network_topology_strategy::make_replication_map(ta
 //    initial_tablets = max(nr_shards_in(dc) / RF_in(dc) for dc in datacenters)
 //
 
-static unsigned calculate_initial_tablets_from_topology(const schema& s, token_metadata_ptr tm, const std::unordered_map<sstring, size_t>& rf) {
-    unsigned initial_tablets = std::numeric_limits<unsigned>::min();
+static future<unsigned> calculate_initial_tablets_from_topology(const schema& s, token_metadata_ptr tm,
+                                                        const std::unordered_map<sstring, size_t>& rf,
+                                                        unsigned initial_scale) {
+    unsigned initial_tablets = 0;
     for (const auto& dc : tm->get_datacenter_token_owners_ips()) {
+        auto& dc_name = dc.first;
         unsigned shards_in_dc = 0;
         unsigned rf_in_dc = 1;
 
         for (const auto& ep : dc.second) {
             const auto* node = tm->get_topology().find_node(ep);
-            if (node != nullptr) {
+            if (node != nullptr && node->is_normal()) {
                 shards_in_dc += node->get_shard_count();
             }
+            co_await coroutine::maybe_yield();
         }
 
         if (auto it = rf.find(dc.first); it != rf.end()) {
             rf_in_dc = it->second;
         }
 
-        unsigned tablets_in_dc = rf_in_dc > 0 ? (shards_in_dc + rf_in_dc - 1) / rf_in_dc : 0;
+        if (rf_in_dc == 0) {
+            continue;
+        }
+
+        uint64_t existing_replicas = 0;
+        for (auto&& [table, tmap] : tm->tablets().all_tables()) {
+            existing_replicas += tmap->tablet_count() * rf_in_dc;
+            co_await coroutine::maybe_yield();
+        }
+
+        // As existing load approaches the limit of tablets per shard, scale down
+        // the initial number of tablets to avoid exceeding the limit.
+        // The tablet scheduler may later fix this by shrinking other tables.
+        // The way it works is that if with an empty cluster initial_scale takes 10%
+        // of shard's budget, then after creating the table the next table
+        // will take 10% of the remaining budget.
+        auto existing_load = double(existing_replicas) / shards_in_dc;
```

Comment:
can we get division by 0 here if we're decommisioning last node in a DC, since we no longer account its shards? maybe 'rf_in_dc == 0' check above is not enough, since to decommission a DC, we recommend removing its entry from the replication properties first (so there might be no entry for this DC in the map), before decommissioning the nodes in it.

Reply:
Currently we don't allow going below RF, so it's not possible. But it's good to be defensive.